# **Aula 3 - Redes Neurais na Prática**

Nas Aulas 1 e 2, construímos um neurônio e uma pequena rede à mão: escolhemos os pesos manualmente e, quando quisemos que a rede aprendesse de verdade, implementamos nós mesmos a regra de atualização de pesos do perceptron -- um exemplo de cada vez.

Isso foi essencial para entender **o que** uma rede neural faz e **por que** ela funciona -- mas não é assim que redes neurais são treinadas na prática. Nesta aula, vamos ver como um **framework** (a biblioteca Keras, sobre o TensorFlow) automatiza esse processo, e por que isso importa: ele permite treinar redes muito maiores, com muito mais dados, sem escrever manualmente a regra de atualização de pesos para cada camada.

Antes de entrarmos no código, vale reforçar a motivação de negócio -- por que, afinal, vale a pena investir nisso.

## 1. Onde as Redes Neurais Fazem Diferença na Prática

**Por que usar uma rede neural em vez de um modelo de Machine Learning "tradicional"?**

Modelos tradicionais de ML (regressão linear/logística, árvores de decisão, random forest, SVM, k-means) continuam sendo ótimas escolhas para muitos problemas -- geralmente são mais rápidos de treinar, mais baratos de rodar em produção e mais fáceis de explicar. A rede neural passa a fazer diferença quando o problema tem características que esses modelos têm dificuldade de capturar:

- **Dados não estruturados**: imagem, áudio, texto livre e vídeo. Modelos tradicionais normalmente exigem que um especialista crie manualmente as variáveis de entrada (feature engineering); redes neurais aprendem essas representações sozinhas, direto dos dados brutos.
- **Relações muito não lineares e com muitas interações entre variáveis**: como vimos na Aula 2, cada camada oculta adicional permite representar combinações mais complexas do que uma equação linear ou uma árvore rasa conseguem.
- **Grandes volumes de dados**: modelos tradicionais tendem a "saturar" o desempenho depois de um certo volume de dados; redes neurais (principalmente as profundas) costumam continuar melhorando à medida que mais dados e mais poder computacional são adicionados.
- **Tarefas de ponta em visão computacional, linguagem natural e fala**, onde redes neurais (CNNs, RNNs/LSTMs, Transformers) hoje são o estado da arte.

Em troca, redes neurais exigem mais dados, mais poder computacional, mais tempo de ajuste, e são mais difíceis de explicar -- o trade-off de explicabilidade que já vimos na tabela comparativa da Aula 2 (redes neurais vs. equações/regras manuais). Ou seja, a escolha entre "ML tradicional" e "rede neural" é uma decisão de custo-benefício, não uma hierarquia onde uma é sempre melhor que a outra.

## 2. Aplicações de Negócio

**12 aplicações de redes neurais em negócios**

Cada uma das aplicações abaixo tem um notebook completo, com dados simulados, código executável e discussão de resultados, na pasta [`aplicacoes-de-negocio/`](aplicacoes-de-negocio/):

1. **Detecção de fraude em transações financeiras** (FFNN) -- [`01-deteccao-fraude.ipynb`](aplicacoes-de-negocio/01-deteccao-fraude.ipynb)
2. **Sistemas de recomendação** (filtragem colaborativa) -- [`02-sistema-recomendacao.ipynb`](aplicacoes-de-negocio/02-sistema-recomendacao.ipynb)
3. **Previsão de demanda na cadeia de suprimentos** -- [`03-previsao-demanda.ipynb`](aplicacoes-de-negocio/03-previsao-demanda.ipynb)
4. **Chatbots e atendimento ao cliente** (análise de sentimento) -- [`04-chatbot-atendimento.ipynb`](aplicacoes-de-negocio/04-chatbot-atendimento.ipynb)
5. **Visão computacional no varejo** (CNN) -- [`05-visao-computacional-varejo.ipynb`](aplicacoes-de-negocio/05-visao-computacional-varejo.ipynb)
6. **Diagnóstico por imagem médica** (CNN) -- [`06-diagnostico-imagem-medica.ipynb`](aplicacoes-de-negocio/06-diagnostico-imagem-medica.ipynb)
7. **Manutenção preditiva industrial** -- [`07-manutencao-preditiva.ipynb`](aplicacoes-de-negocio/07-manutencao-preditiva.ipynb)
8. **Precificação dinâmica** -- [`08-precificacao-dinamica.ipynb`](aplicacoes-de-negocio/08-precificacao-dinamica.ipynb)
9. **Score de crédito e análise de risco** -- [`09-score-credito.ipynb`](aplicacoes-de-negocio/09-score-credito.ipynb)
10. **Geração de conteúdo e marketing** (GANs) -- [`10-geracao-conteudo-marketing.ipynb`](aplicacoes-de-negocio/10-geracao-conteudo-marketing.ipynb)
11. **Prevenção à lavagem de dinheiro (AML)** (Autoencoder) -- [`11-prevencao-lavagem-dinheiro.ipynb`](aplicacoes-de-negocio/11-prevencao-lavagem-dinheiro.ipynb)
12. **Gestão de pessoas (People Analytics)** -- [`12-people-analytics-turnover.ipynb`](aplicacoes-de-negocio/12-people-analytics-turnover.ipynb)

Como bônus, a mesma pasta ainda tem um 13º notebook -- [`13-mini-llm-poucos-textos.ipynb`](aplicacoes-de-negocio/13-mini-llm-poucos-textos.ipynb) -- sobre a construção de uma mini LLM a partir de poucos textos.

Repare que a maioria dessas aplicações combina os tipos de rede que vamos ver na Aula 4 (CNN, RNN/LSTM, GAN, Autoencoder) com o problema específico do negócio.

### Tipos de função de ativação

Já conhecemos três funções de ativação: o degrau (Aula 1, perceptron), e a ReLU e a sigmoid (Aula 2). A tabela abaixo situa essas três ao lado de outras comuns em redes mais profundas:

| **Função de Ativação** | **Intervalo de Saída** | **Uso Principal** | **Vantagens** | **Desvantagens** |
|----------------|----------------|----------------|----------------|----------------|
| **Degrau (Step Function)** | {0,1} | Perceptron simples | Fácil de entender | Apenas resolve problemas linearmente separáveis |
| **Sigmoid** | (0,1) | Classificação binária | Interpretação probabilística | Vanishing gradient |
| **Tanh** | (-1,1) | Classificação binária | Melhor que sigmoid, saída centrada | Ainda sofre de vanishing gradient |
| **ReLU** | [0,∞) | Redes profundas | Resolve vanishing gradient | Dying ReLU |
| **Leaky ReLU** | (-∞,∞) | Redes profundas | Evita dying ReLU | Requer ajuste do coeficiente |
| **Softmax** | (0,1) | Classificação multiclasse | Transforma em probabilidades | Computacionalmente caro |

## 3. Por que usar um framework?

Na Aula 2, cada peso da rede foi escolhido por nós manualmente. Na demonstração de quebra de simetria, chegamos a *treinar* pesos de verdade, mas usando a regra do perceptron -- que só sabe ajustar os pesos de **uma única camada**, comparando a saída diretamente com o rótulo.

Numa rede com camada oculta, o problema fica mais difícil: como saber o quanto cada peso da camada oculta contribuiu para o erro final, se o erro só é calculado na camada de saída? A resposta é o **backpropagation** (RUMELHART; HINTON; WILLIAMS, 1986): um algoritmo que propaga o erro da camada de saída de volta para as camadas anteriores, calculando o quanto cada peso -- de cada camada -- deveria mudar. É, na prática, a mesma ideia da regra do perceptron (ajustar o peso proporcionalmente ao erro), só que generalizada matematicamente para redes com várias camadas.

Implementar backpropagation à mão para uma rede grande seria trabalhoso e propenso a erros. É exatamente isso que frameworks como **Keras** (sobre o **TensorFlow**) automatizam: você descreve a arquitetura (camadas, neurônios, ativações) e o framework calcula e aplica as atualizações de peso de cada camada automaticamente, a cada época de treino.

## 4. Revisitando o e-mail de spam com Keras

Vamos usar exatamente o mesmo problema e a mesma arquitetura da Aula 2 (3 entradas → 3 neurônios ocultos com ReLU → 1 neurônio de saída com sigmoid) -- mas, desta vez, em vez de escolher os pesos à mão, vamos deixar o Keras aprendê-los a partir de dados, usando backpropagation.

Para isso, também precisamos de mais dados do que os 10 e-mails da Aula 1: com poucos exemplos, qualquer rede "decora" o treino em vez de aprender um padrão generalizável. Vamos gerar uma base sintética maior (600 e-mails), com as mesmas 3 características (links, linguagem suspeita, confiabilidade do remetente), e separar uma parte para teste -- e-mails que a rede nunca viu durante o treino -- para medir a acurácia de forma honesta.

In [1]:
import numpy as np
from sklearn.model_selection import train_test_split

seed = 42
np.random.seed(seed)

n_por_classe = 300

# E-mails de spam: muitos links, linguagem suspeita, remetente pouco confiavel
spam = np.column_stack([
    np.clip(np.random.normal(7, 2, n_por_classe), 0, 10),
    np.clip(np.random.normal(7, 2, n_por_classe), 0, 10),
    np.clip(np.random.normal(2, 1.5, n_por_classe), 0, 10),
])

# E-mails legitimos: poucos links, linguagem normal, remetente confiavel
nao_spam = np.column_stack([
    np.clip(np.random.normal(2, 1.5, n_por_classe), 0, 10),
    np.clip(np.random.normal(2, 1.5, n_por_classe), 0, 10),
    np.clip(np.random.normal(7, 2, n_por_classe), 0, 10),
])

X = np.vstack([spam, nao_spam])
y = np.concatenate([np.ones(n_por_classe), np.zeros(n_por_classe)])

X_treino, X_teste, y_treino, y_teste = train_test_split(
    X, y, test_size=0.2, random_state=seed, stratify=y
)

print('Total de e-mails simulados:', len(X))
print('Exemplos de treino:', len(X_treino))
print('Exemplos de teste:', len(X_teste))

Total de e-mails simulados: 600
Exemplos de treino: 480
Exemplos de teste: 120


In [2]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Input

tf.random.set_seed(seed)

modelo = Sequential()
modelo.add(Input(shape=(3,)))
modelo.add(Dense(3, activation='relu', name='CamadaOculta'))
modelo.add(Dense(1, activation='sigmoid', name='CamadaSaida'))

modelo.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

print("Treinando a rede neural...")
historico = modelo.fit(X_treino, y_treino, epochs=100, verbose=0)

perda_teste, acuracia_teste = modelo.evaluate(X_teste, y_teste, verbose=0)
print(f"Acuracia no conjunto de teste (e-mails nunca vistos no treino): {acuracia_teste:.2%}")

Treinando a rede neural...


Acuracia no conjunto de teste (e-mails nunca vistos no treino): 98.33%


### Explicação do Código

Cada peça do Keras corresponde a algo que já vimos nas Aulas 1 e 2:

- **`Sequential`**: uma pilha de camadas, o mesmo fluxo entrada → oculta → saída que já usamos.
- **`Dense(3, activation='relu')`**: a camada oculta com 3 neurônios e ReLU, igual à da Aula 2 -- só que agora os pesos começam **aleatórios** (a quebra de simetria da Aula 2, aplicada automaticamente) e são ajustados durante o treino, em vez de fixados à mão.
- **`Dense(1, activation='sigmoid')`**: a camada de saída, igual à da Aula 2.
- **`compile(optimizer='adam', loss='binary_crossentropy')`**: escolhe o algoritmo que ajusta os pesos a cada passo (uma versão mais sofisticada de `peso = peso + taxa_aprendizado * erro * entrada`, a mesma regra da Aula 1) e a forma de medir o erro para um problema de classificação binária.
- **`fit(X_treino, y_treino, epochs=100)`**: roda o backpropagation por 100 épocas -- o mesmo conceito de "época" da Aula 1.
- **`evaluate` / `predict`**: a mesma inferência que já calculamos manualmente na Aula 2, agora automática.

In [3]:
email_exemplo = np.array([[8.0, 7.0, 2.0]])  # o mesmo e-mail de exemplo desde a Aula 1
probabilidade = modelo.predict(email_exemplo, verbose=0)
print(f"Probabilidade de spam para o e-mail de exemplo [links=8, linguagem_suspeita=7, confiabilidade=2]: {probabilidade[0][0]:.2%}")

Probabilidade de spam para o e-mail de exemplo [links=8, linguagem_suspeita=7, confiabilidade=2]: 99.03%


Na Aula 2, com pesos escolhidos manualmente, esse mesmo e-mail recebeu uma probabilidade de spam de aproximadamente 83%. Agora, com pesos **aprendidos** a partir de 480 exemplos de treino via backpropagation, a rede chega a uma conclusão semelhante -- ambos os caminhos (pesos manuais bem escolhidos e pesos aprendidos de dados reais) convergem para a mesma direção qualitativa, o que é uma boa checagem de sanidade: a intuição que construímos à mão nas Aulas 1 e 2 é consistente com o que a rede aprende sozinha.

## Fechamento: o que vem a seguir

Agora que sabemos treinar uma rede "feedforward" simples com um framework de verdade, a pergunta natural é: essa mesma arquitetura (camadas densas, uma após a outra) serve para qualquer tipo de dado?

A resposta é não. Imagens, sequências temporais e texto têm estruturas próprias (vizinhança espacial, ordem temporal) que uma rede densa comum ignora. Na Aula 4, vamos conhecer arquiteturas especializadas -- CNNs, RNNs, LSTMs, GANs e Autoencoders -- cada uma criada para explorar a estrutura específica de um tipo de dado, e ver por que cada uma resolve uma limitação da anterior.

# **Referências**

RUMELHART, D. E.; HINTON, G. E.; WILLIAMS, R. J. Learning representations by back-propagating errors. **Nature**, [s. l.], v. 323, n. 6088, p. 533-536, 1986.